Scripts/generate-test-config.sh

```diff
@@ -1,14 +1,269 @@
+  log_error $*
   exit 1
+}
+
+# Usage & help
+
+function print_usage {
+  echo "usage: ${SYNOPSIS}"
+}
+
+function print_help {
+  usage=$(print_usage)
+  cat <<EOF
+${usage}
+
+SYNOPSIS
+    ${SYNOPSIS}
+
+$SCRIPT_NAME generates a test configuration file for AWS SDK and
+Amplify integration tests.
+
+REQUIREMENTS
+    $SCRIPT_NAME requires the following utilities to be installed
+    and available on your PATH:
+        - Python - v3.7 or higher, accessible as 'python3'
+          (https://www.python.org/)
+    
+    If you use the -a option to assume the test execution role prior to
+    generating the test configuration file, you must also have the following
+    utilities installed:
+        - jq - A JSON parsing utility
+          (https://stedolan.github.io/jq/)
+        - AWS CLI - for retrieving secret values from STS
+          (https://docs.aws.amazon.com/cli/latest/userguide/cli-chap-install.html)
+
+OPTIONS:
+    -a
+        Assume the test execution role before generating the test configuration
+        file. Requires your environment to have exported credentials for a role
+        or user that has 'sts:AssumeRole' permissions for the integration test
+        execution role. If not specified, temporary role credentials must be
+        exported into the environment variables 'AWS_ACCESS_KEY_ID',
+        'AWS_SECRET_ACCESS_KEY', and 'AWS_SESSION_TOKEN'.
+
+    -p <android|ios>
+        Generate configuration file for platform="<val>"
+
+    -r <region>
+        Look for resources in <region>. If not present, use the value of the
+        environment variable AWS_DEFAULT_REGION. Either the environment
+        variable or this flag must be specified.
+
+    STANDARD OPTIONS:
+    -h
+        Print this help message
+    -q
+        Quiet mode, only errors will be logged
+    -v
+        Verbose output. Specify up to 2 times for more debugging output
+EOF
+}
+
+
+##################################################
+# PARSE ARGUMENTS
+
+assume_role=""
+platform=""
+region=${AWS_DEFAULT_REGION}
+
+while getopts "ap:qvh" optchar ; do
+  case "$optchar" in
+    a)
+      assume_role=1
+      ;;
+    p)
+      platform=$OPTARG
+      ;;
+    r)
+      region=$OPTARG
+      ;;
+
+    q)
+      LOG_LEVEL=-1
+      ;;
+    v)
+      LOG_LEVEL+=1
+      ;;
+    h)
+      print_help
+      exit 0
+      ;;
+    *)
+      print_usage >&2
+      exit 1
+      ;;
+  esac
+done
+
+
+##################################################
+# VALIDATIONS AND ASSIGNMENTS
+
+cmd_quiet_flag="--quiet"
+if [[ $LOG_LEVEL -gt 1 ]] ; then
+  cmd_quiet_flag=""
+fi
+
+if [[ -z $platform ]] ; then
+  die "'platform' not specified"
+fi
+
+if [[ -z $region ]] ; then
+  die "'region' not specified"
+fi
+
+python_cmd=$(which python3)
+if [[ -z $python_cmd ]] ; then
+  die "Cannot find 'python3' in PATH"
+fi
+
+if [[ $assume_role ]] ; then
+  aws_cli_cmd=$(which aws)
+  if [[ -z $aws_cli_cmd ]] ; then
+    die "Cannot find 'aws' in PATH"
+  fi
+
+  jq_cmd=$(which jq)
```

Comment:
Let's say that a naked call to `jq` will resolve to `/on/your/path/jq`. Then `$(which jq)` would _also_ result in a call to `/on/your/path/jq`. So, you may not gain a lot from storing the path, above just doing `jq`, later.

If you want to be careful about the script path, you could clobber `PATH` at the beginning of your script, and then reconstruct it with "safe" directories, where you believe you will find well-known and un-compromised utility software (`"/usr/bin:/bin:/usr/sbin:/sbin"`, for example.)

Debian stable used to be even more conservative than that, and not include `/usr/bin` in `root`'s default `$PATH`, I believe.

Reply:
Thanks for the callout. That's a leftover from me originally specifying paths to executable dependencies as hardcoded values rather than auto-discovering them with `which`. Funny thing is, I didn't even use `$jq_cmd` and `$aws_cli_cmd` down where I actually invoke it, so the assignment wasn't going to be useful anyway. Sheesh.

I'm not going to be overly paranoid about the path, since these are going to be invoked interactively by developers who may have installed into non-traditional paths (e.g., python in pyenv shim folders, homebrew-managed packages, etc).